### PR TITLE
Allow random state variables to be synced across TP.

### PR DESCRIPTION
**Pull request by thomasw21**
Upstream change from https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/276.

Essentially in BigScience, we noticed states going out of sync, in particular `torch.rng_state`. It was mainly due to the data loader using rng_state to generate seeds for workers. This causes the main process to consume a RNG state,  while other TP ranks didn't. This cause a change in RNG state between `tp_rank = 0` and `tp_rank > 0`, which some layers relie on (for example, `nn.Dropout`). More details in https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/276

### CODE REVIEW

**Comment by sbhavani** [comment]
@thomasw21 Thanks for your contribution on syncing RNG states across TP! As part of our new year cleanup, we're closing PRs opened before 2025 since the codebase has evolved significantly.

The RNG management has been restructured, though the specific DataLoader fix may still be relevant @NVIDIA/megatron-oncall 

If you'd like to revisit this, please feel free to reopen or submit a new PR against the latest main branch. Thanks for your understanding!
